File: 06_longest_common_prefix.ts
// optimized solution
function longestCommonPrefix(strs: string[]): string {
    if (strs.length === 0) return "";

    let prefix = strs[0];

    for (let i = 0; i < strs.length; i++) {

        while (strs[i].indexOf(prefix) != 0) {
            prefix = prefix.slice(0, prefix.length - 1)

            console.log("prefix", prefix)

            if (prefix === "") return ""
        }
    }

    return prefix
}

const result_06 = longestCommonPrefix(["flight","flow","flower"])            // "fl"
// const result_06 = longestCommonPrefix(["dog","racecar","car"])               // ""
// const result_06 = longestCommonPrefix(["daf","dax","dat"])                   // "da"
// const result_06 = longestCommonPrefix(["d","d","d","d"])                     // "d"
// const result_06 = longestCommonPrefix(["d","d","d"])                         // "d"
// const result_06 = longestCommonPrefix(["a"])                                 // "a"
// const result_06 = longestCommonPrefix(["flower","flower","flower","flower"]) // "flower"
// const result_06 = longestCommonPrefix(["cir","car"])                         // "c"

console.log("result: ", result_06)

File: 06_longest_common_prefix.rs
struct Solution {}

impl Solution {
    pub fn longest_common_prefix(strs: Vec<&str>) -> &str {
        println!("{}", strs.len());

        let mut prefix = strs[0];

        for i in 0..strs.len() {
            while strs[i].find(prefix) != Some(0) {
                prefix = &prefix[0..prefix.len() - 1];

                if prefix == "" { return "" }
            }
        }
        
        if strs.len() == 0 { return "" };
        
        return prefix
    }
}

fn main() {
    let strs: Vec<&str> = vec!["flower", "flow", "flight"];
    let result: &str = Solution::longest_common_prefix(strs);
    println!("{}", result);
}